### PR TITLE
Auto-notes for map extras

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Interface "Auto-notes for map extras"```

#### Purpose of change

Allow auto-notes for map extras.

#### Describe the solution

- added new `map_extra` class and implemented its loading from json (to allow configurable descriptions in auto-notes, overriding in mods, etc);
- added option for automatic creation of notes of revealed map extras;
- fixed a bug with display of text with color tags in preview window;
- bumped maximum overmap note length 10x times (to 450 characters).

#### Describe alternatives you've considered

Implement additional map view overlay for map extras.

Map extras are also very similar to mapgen functions and nested mapgens - we should consider merging these entities together in the future to de-duplicate code.

#### Screenshots

_New auto-notes_

image

_New options_
image

### CODE REVIEW

**Comment by Inglonias** [comment]
Woooooo

File: src/map_extras.cpp [hunk 1]
```diff
@@ -2298,25 +2318,64 @@ FunctionMap builtin_functions = {
...
+    mandatory( jo, was_loaded, "description", description );
+    mandatory( jo, was_loaded, "function", function );
+    function_pointer = MapExtras::get_function( function );
+    if( function_pointer == nullptr ) {
+        debugmsg( "invalid map extra function (%s) defined for map extra (%s)", function, id.str() );
+    }
+    optional( jo, was_loaded, "sym", symbol, unicode_codepoint_from_symbol_reader, NULL_UNICODE );
+    color = jo.has_member( "color" ) ? color_from_string( jo.get_string( "color" ) ) : c_white;
+    optional( jo, was_loaded, "autonote", autonote, false );
 } // namespace MapExtras
```
Comment (ZhilkinSerg) [inline src/map_extras.cpp]:
```suggestion
}
```

File: src/map_extras.cpp [hunk 2]
```diff
@@ -2298,25 +2318,64 @@ FunctionMap builtin_functions = {
...
 {
     return builtin_functions;
 }
 
+void load( JsonObject &jo, const std::string &src )
+{
+    extras.load( jo, src );
+}
+
+}
```
Comment (ZhilkinSerg) [inline src/map_extras.cpp]:
```suggestion
} // namespace MapExtras
```

File: src/map_extras.cpp [hunk 3]
```diff
@@ -41,9 +42,25 @@
...
+generic_factory<map_extra> extras( "map extra" );
+
+}
+
+/** @relates string_id */
+template<>
+const map_extra &string_id<map_extra>::obj() const
+{
+    return extras.obj( *this );
+}
```
Comment (ZhilkinSerg) [inline src/map_extras.cpp]:
```suggestion
} // namespace
```

File: src/map_extras.cpp [hunk 4]
```diff
@@ -41,9 +42,25 @@
...
+#include "options.h"
 
 class npc_template;
 
+namespace
+{
+
+generic_factory<map_extra> extras( "map extra" );
+
+}
```
Comment (ZhilkinSerg) [inline src/map_extras.cpp]:
```suggestion
} // namespace
```

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild.

**Pull request merged**